### PR TITLE
QtFRED: Custom Warp Dialog

### PR DESCRIPTION
**Pull request by TheForce172**
Not much to say on this one. Functions almost exactly the same as the oldFRED one. 

Only changes are not showing lines irrelevant to the selected warp type and additional warnings if game sounds don't not exist.

Also visually shows the Hyperspace warp type as "Star Wars" for clearer useability but does not change the backend.

### CODE REVIEW

**Review by Goober5000 — CHANGES_REQUESTED** [review]
I built the code and took a look.  Couple comments...
1. The "All of these parameters are optional..." explanatory text should be displayed on the dialog as a usability aid.
2. In standard FRED, the dialog title displays "Warp-In Parameters" or "Warp-Out Parameters" depending on what is being edited.  The qtFRED version should include "Warp-In" or "Warp-Out" as well.
3. The "qtFRED v24.1.0" suffix is only really needed on the application title bar, not the title bar of every dialog.
4. "Acceleration" is misspelled as "Aceleration"
5. Changing "Hyperspace" to "Star Wars" makes sense, since many of the other types are named after their respective sci-fi universes.

As far as hiding inapplicable fields, rather than displaying but disabling them... I like the idea of making the fields visible but disabled so that the FREDder can see what can be done by changing the warp type, but that could be a personal preference thing.  However, the hiding of fields does cause occasional layout problems.  This can be seen most easily when switching between Homeworld and Star Wars.

**Comment by TheForce172** [comment]
Points 1, 2 and 4 addressed, also switched to disabling elements rather than hiding them. Point 3 is beyond the scope of this PR as that is done by either QT or the Core QtFred code it's not set on a per dialog option. Glad you approve on point 5.

File: qtfred/src/ui/dialogs/ShipEditor/ShipEditorDialog.cpp [hunk 1]
```diff
@@ -879,7 +879,8 @@ void ShipEditorDialog::on_restrictArrivalPathsButton_clicked()
 }
 void ShipEditorDialog::on_customWarpinButton_clicked()
 {
-	// TODO:: Custom warp Dialog
+	auto dialog = new dialogs::ShipCustomWarpDialog(this, _viewport, false);
```
Comment (Goober5000) [inline qtfred/src/ui/dialogs/ShipEditor/ShipEditorDialog.cpp:882]: Is this going to cause a memory leak?  What happens to the `dialog` pointer when it goes out of scope?

File: qtfred/src/ui/dialogs/ShipEditor/ShipEditorDialog.cpp [hunk 2]
```diff
@@ -888,7 +889,9 @@ void ShipEditorDialog::on_restrictDeparturePathsButton_clicked()
 	dialog->show();
 }
 void ShipEditorDialog::on_customWarpoutButton_clicked()
-{ // TODO:: Custom warp Dialog
+{
+	auto dialog = new dialogs::ShipCustomWarpDialog(this, _viewport, true);
```
Comment (Goober5000) [inline qtfred/src/ui/dialogs/ShipEditor/ShipEditorDialog.cpp:893]: Same here

File: qtfred/ui/ShipCustomWarpDialog.ui [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+   <string>Edit Parameters</string>
+  </property>
+  <property name="modal">
+   <bool>true</bool>
+  </property>
+  <layout class="QVBoxLayout" name="layout">
+   <item>
+    <widget class="QLabel" name="label">
+     <property name="text">
+      <string>Note: All these parameters are optional.</string>
```
Comment (Goober5000) [inline qtfred/ui/ShipCustomWarpDialog.ui:26]: I meant that the full usability text that is used in FRED should also be used in qtFRED:
> All of these parameters are optional, and a mission designer will usually not need to override more than one or two.  These parameters are the same as those found in ships.tbl.

File: qtfred/src/ui/dialogs/ShipEditor/ShipCustomWarpDialog.cpp [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+	connect(ui->lineEditAnim, &QLineEdit::editingFinished, this, &ShipCustomWarpDialog::animChanged);
+	connect(ui->doubleSpinBoxPlayerSpeed,
+		QOverload<double>::of(&QDoubleSpinBox::valueChanged),
+		_model.get(),
+		&ShipCustomWarpDialogModel::setPlayerSpeed);
+
+	if (departure) {
+		this->setWindowTitle("Edit Warp-Out Parameters");
+	} else {
+		this->setWindowTitle("Edit Warp-in Parameters");
```
Comment (Goober5000) [inline qtfred/src/ui/dialogs/ShipEditor/ShipCustomWarpDialog.cpp:60]: The "I" in "Warp-in" should be capitalized for consistency

File: qtfred/src/ui/dialogs/ShipEditor/ShipCustomWarpDialog.cpp [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+			accept();
+			return;
+		}
+		if (button == DialogButton::No) {
+			_model->reject();
+		}
+	}
+	QDialog::closeEvent(e);
+}
+void ShipCustomWarpDialog::updateUI(const bool firstrun)
```
Comment (Goober5000) [inline qtfred/src/ui/dialogs/ShipEditor/ShipCustomWarpDialog.cpp:91]: In this function, rather than using string comparisons to the warp type name, it would be better to compare the warp type index (e.g. `i` or `ui->comboBoxType->currentIndex()`) to one of the `WT_` definitions in shipfx.h.  This makes the code more maintainable.  (It's also slightly faster.)

File: qtfred/src/ui/dialogs/ShipEditor/ShipCustomWarpDialog.cpp [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+
+	if (ui->comboBoxType->itemText(ui->comboBoxType->currentIndex()) == "Star Wars") {
+		ui->doubleSpinBoxExponent->setEnabled(true);
+		ui->labelExponent->setEnabled(true);
+	} else {
+		ui->doubleSpinBoxExponent->setEnabled(false);
+		ui->labelExponent->setEnabled(false);
+	}
+}
+void ShipCustomWarpDialog::rejectHandler()
```
Comment (Goober5000) [inline qtfred/src/ui/dialogs/ShipEditor/ShipCustomWarpDialog.cpp:154]: This seems like a common enough pattern that it should be in the superclass or in a utility function somewhere, though maybe that's beyond the scope of this PR.

File: qtfred/src/ui/dialogs/ShipEditor/ShipCustomWarpDialog.cpp [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+			QDialog::reject();
+		}
+	} else {
+		_model->reject();
+		QDialog::reject();
+	}
+}
+void ShipCustomWarpDialog::startSoundChanged()
+{
+	// String wrangling reqired in order to avoid crashes when directly converting from Qstring to std::string on some
```
Comment (Goober5000) [inline qtfred/src/ui/dialogs/ShipEditor/ShipCustomWarpDialog.cpp:181]: The string wrangling should definitely be in a utility function.  If you want to do that in a different PR that's fine; but if string wrangling is necessary, then there should be qtFRED-wide utility functions for converting from Qstring to std::string and vice versa.

File: qtfred/src/mission/dialogs/ShipEditor/ShipCustomWarpDialogModel.cpp [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+	if (!_m_anim.empty()) {
+		strcpy_s(params.anim, _m_anim.c_str());
+	}
+	params.supercap_warp_physics = _m_supercap_warp_physics;
+	if (_m_departure && _m_player_warpout_speed) {
+		params.warpout_player_speed = _m_player_warpout_speed;
+	}
+	int index = find_or_add_warp_params(params);
+
+	for (object* objp = GET_FIRST(&obj_used_list); objp != END_OF_LIST(&obj_used_list); objp = GET_NEXT(objp)) {
```
Comment (Goober5000) [inline qtfred/src/mission/dialogs/ShipEditor/ShipCustomWarpDialogModel.cpp:64]: You can use `for (object* objp: list_range(&obj_used_list))` here and in similar loops if you want.

**Review by Goober5000 — CHANGES_REQUESTED** [review]
Checked the code and noticed a few things, mostly minor.

File: qtfred/src/ui/dialogs/ShipEditor/ShipEditorDialog.cpp (hunk 2 above)
Comment (TheForce172) [reply qtfred/src/ui/dialogs/ShipEditor/ShipEditorDialog.cpp:893]: See above

File: qtfred/src/ui/dialogs/ShipEditor/ShipCustomWarpDialog.cpp (hunk 7 above)
Comment (TheForce172) [reply qtfred/src/ui/dialogs/ShipEditor/ShipCustomWarpDialog.cpp:181]: Agreed, but that should probably be done in a separate PR as there are many places in existing dialogs it should be put.

File: qtfred/ui/ShipCustomWarpDialog.ui (hunk 3 above)
Comment (TheForce172) [reply qtfred/ui/ShipCustomWarpDialog.ui:26]: Not particularly happy with this one as the text seems rather verbose and unnecessary but don't feel strongly enough to fight you on this.

File: qtfred/src/ui/dialogs/ShipEditor/ShipEditorDialog.cpp (hunk 1 above)
Comment (Goober5000) [reply qtfred/src/ui/dialogs/ShipEditor/ShipEditorDialog.cpp:882]: I did a bit of searching around.  Basically, there are a few options:
1. Display a modal dialog with `exec()` rather than `show()` and then delete the pointer immediately afterward
2. Wait until the program exits at which point the child dialogs will be deleted automatically (requires the dialog to be created with a parent)
3. Use a global variable so there is a single dialog that is hidden or shown as needed, rather than being created each time
4. Set the delete-on-close attribute via `diag->setAttribute(Qt::WA_DeleteOnClose);`

**Review by Goober5000 — APPROVED** [review]

File: qtfred/src/ui/dialogs/ShipEditor/ShipEditorDialog.cpp (hunk 1 above)
Comment (TheForce172) [reply qtfred/src/ui/dialogs/ShipEditor/ShipEditorDialog.cpp:882]: Option 3 is what I attempted with the other dialogs and prove a bit kludge and difficult to manage making sure it had the right data. Of these I'm leaning towards option 4 being best for our use cases.

Comment (Goober5000) [reply qtfred/src/ui/dialogs/ShipEditor/ShipEditorDialog.cpp:882]: Agreed, option 4 sounds pretty good

**Comment by Goober5000** [comment]
I've created separate issues for the outside-scope items mentioned, so I'm happy to merge this now.
